src/pubsub.c

```diff
@@ -695,25 +734,28 @@ NULL
     }
 }
 
-void channelList(client *c, sds pat, dict *pubsub_channels) {
-    dictIterator *di = dictGetIterator(pubsub_channels);
-    dictEntry *de;
+void channelList(client *c, sds pat, dict **pubsub_channels, int is_sharded) {
     long mblen = 0;
     void *replylen;
+    unsigned int slot_cnt = is_sharded ? CLUSTER_SLOTS : 1;
 
     replylen = addReplyDeferredLen(c);
-    while((de = dictNext(di)) != NULL) {
-        robj *cobj = dictGetKey(de);
-        sds channel = cobj->ptr;
-
-        if (!pat || stringmatchlen(pat, sdslen(pat),
-                                   channel, sdslen(channel),0))
-        {
-            addReplyBulk(c,cobj);
-            mblen++;
+    for (unsigned int i = 0; i < slot_cnt; i++) {
+        dictIterator *di = dictGetIterator(pubsub_channels[i]);
```

Comment:
we should first check if `pubsub_channels[i]` is `NULL`, in case the null pointer exception.

and it's better add a test case about it.